src/binary_trees.rs

```diff
@@ -13,26 +13,42 @@ use typed_arena::Arena;
 use rayon::prelude::*;
 
 struct Tree<'a> {
-    l: Option<&'a Tree<'a>>,
-    r: Option<&'a Tree<'a>>,
-    i: i32,
+    children: Option<(&'a Tree<'a>, &'a Tree<'a>)>,
```

Comment:
A node has either two children or none, this saves half the null-checks in the matching code below.